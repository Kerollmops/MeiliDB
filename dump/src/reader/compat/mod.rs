// pub mod v2;
// pub mod v3;
// pub mod v4;
use crate::Result;

use self::{
    v4_to_v5::CompatV4ToV5,
    v5_to_v6::{CompatIndexV5ToV6, CompatV5ToV6},
};

use super::{
    v5::V5Reader,
    v6::{self, V6IndexReader, V6Reader},
    Document, UpdateFile,
};

pub mod v2_to_v3;
pub mod v3_to_v4;
pub mod v4_to_v5;
pub mod v5_to_v6;

pub enum Compat {
    Current(V6Reader),
    Compat(CompatV5ToV6),
}

impl Compat {
    pub fn version(&self) -> crate::Version {
        match self {
            Compat::Current(current) => current.version(),
            Compat::Compat(compat) => compat.version(),
        }
    }

    pub fn date(&self) -> Option<time::OffsetDateTime> {
        match self {
            Compat::Current(current) => current.date(),
            Compat::Compat(compat) => compat.date(),
        }
    }

    pub fn instance_uid(&self) -> Result<Option<uuid::Uuid>> {
        match self {
            Compat::Current(current) => current.instance_uid(),
            Compat::Compat(compat) => compat.instance_uid(),
        }
    }

    pub fn indexes(&self) -> Result<Box<dyn Iterator<Item = Result<CompatIndex>> + '_>> {
        match self {
            Compat::Current(current) => {
                let indexes = Box::new(current.indexes()?.map(|res| res.map(CompatIndex::from)))
                    as Box<dyn Iterator<Item = Result<CompatIndex>> + '_>;
                Ok(indexes)
            }
            Compat::Compat(compat) => {
                let indexes = Box::new(compat.indexes()?.map(|res| res.map(CompatIndex::from)))
                    as Box<dyn Iterator<Item = Result<CompatIndex>> + '_>;
                Ok(indexes)
            }
        }
    }

    pub fn tasks(
        &mut self,
    ) -> Box<dyn Iterator<Item = Result<(v6::Task, Option<Box<UpdateFile>>)>> + '_> {
        match self {
            Compat::Current(current) => current.tasks(),
            Compat::Compat(compat) => compat.tasks(),
        }
    }

    pub fn keys(&mut self) -> Box<dyn Iterator<Item = Result<v6::Key>> + '_> {
        match self {
            Compat::Current(current) => current.keys(),
            Compat::Compat(compat) => compat.keys(),
        }
    }
}

impl From<V6Reader> for Compat {
    fn from(value: V6Reader) -> Self {
        Compat::Current(value)
    }
}

impl From<CompatV5ToV6> for Compat {
    fn from(value: CompatV5ToV6) -> Self {
        Compat::Compat(value)
    }
}

impl From<V5Reader> for Compat {
    fn from(value: V5Reader) -> Self {
        Compat::Compat(value.to_v6())
    }
}

impl From<CompatV4ToV5> for Compat {
    fn from(value: CompatV4ToV5) -> Self {
        Compat::Compat(value.to_v6())
    }
}

pub enum CompatIndex {
    Current(v6::V6IndexReader),
    Compat(CompatIndexV5ToV6),
}

impl CompatIndex {
    pub fn new_v6(v6: v6::V6IndexReader) -> CompatIndex {
        CompatIndex::Current(v6)
    }

    pub fn metadata(&self) -> &crate::IndexMetadata {
        match self {
            CompatIndex::Current(v6) => v6.metadata(),
            CompatIndex::Compat(compat) => compat.metadata(),
        }
    }

    pub fn documents(&mut self) -> Result<Box<dyn Iterator<Item = Result<Document>> + '_>> {
        match self {
            CompatIndex::Current(v6) => v6
                .documents()
                .map(|iter| Box::new(iter) as Box<dyn Iterator<Item = Result<Document>> + '_>),
            CompatIndex::Compat(compat) => compat
                .documents()
                .map(|iter| Box::new(iter) as Box<dyn Iterator<Item = Result<Document>> + '_>),
        }
    }

    pub fn settings(&mut self) -> Result<v6::Settings<v6::Checked>> {
        match self {
            CompatIndex::Current(v6) => v6.settings(),
            CompatIndex::Compat(compat) => compat.settings(),
        }
    }
}

impl From<V6IndexReader> for CompatIndex {
    fn from(value: V6IndexReader) -> Self {
        CompatIndex::Current(value)
    }
}

impl From<CompatIndexV5ToV6> for CompatIndex {
    fn from(value: CompatIndexV5ToV6) -> Self {
        CompatIndex::Compat(value)
    }
}
