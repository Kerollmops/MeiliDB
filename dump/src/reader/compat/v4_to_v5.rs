use crate::reader::{v4, v5, DumpReader, IndexReader};
use crate::Result;

use super::v3_to_v4::{CompatIndexV3ToV4, CompatV3ToV4};
use super::v5_to_v6::CompatV5ToV6;

pub enum CompatV4ToV5 {
    V4(v4::V4Reader),
    Compat(CompatV3ToV4),
}

impl CompatV4ToV5 {
    pub fn new(v4: v4::V4Reader) -> CompatV4ToV5 {
        CompatV4ToV5::V4(v4)
    }

    pub fn to_v6(self) -> CompatV5ToV6 {
        CompatV5ToV6::Compat(self)
    }

    pub fn version(&self) -> crate::Version {
        match self {
            CompatV4ToV5::V4(v4) => v4.version(),
            CompatV4ToV5::Compat(compat) => compat.version(),
        }
    }

    pub fn date(&self) -> Option<time::OffsetDateTime> {
        match self {
            CompatV4ToV5::V4(v4) => v4.date(),
            CompatV4ToV5::Compat(compat) => compat.date(),
        }
    }

    pub fn instance_uid(&self) -> Result<Option<uuid::Uuid>> {
        match self {
            CompatV4ToV5::V4(v4) => v4.instance_uid(),
            CompatV4ToV5::Compat(compat) => compat.instance_uid(),
        }
    }

    pub fn indexes(&self) -> Result<Box<dyn Iterator<Item = Result<CompatIndexV4ToV5>> + '_>> {
        Ok(match self {
            CompatV4ToV5::V4(v4) => Box::new(
                v4.indexes()?
                    .map(|index| index.map(CompatIndexV4ToV5::from)),
            )
                as Box<dyn Iterator<Item = Result<CompatIndexV4ToV5>> + '_>,

            CompatV4ToV5::Compat(compat) => Box::new(
                compat
                    .indexes()?
                    .map(|index| index.map(CompatIndexV4ToV5::from)),
            )
                as Box<dyn Iterator<Item = Result<CompatIndexV4ToV5>> + '_>,
        })
    }

    pub fn tasks(
        &mut self,
    ) -> Box<dyn Iterator<Item = Result<(v5::Task, Option<v5::UpdateFile>)>> + '_> {
        let tasks = match self {
            CompatV4ToV5::V4(v4) => v4.tasks(),
            CompatV4ToV5::Compat(compat) => compat.tasks(),
        };
        Box::new(tasks.map(|task| {
            task.map(|(task, content_file)| {
                // let task_view: v4::tasks::TaskView = task.into();

                let task = v5::Task {
                    id: task.id,
                    content: match task.content {
                        v4::tasks::TaskContent::DocumentAddition {
                            content_uuid,
                            merge_strategy,
                            primary_key,
                            documents_count,
                            allow_index_creation,
                        } => v5::tasks::TaskContent::DocumentAddition {
                            index_uid: v5::meta::IndexUid(task.index_uid.0),
                            content_uuid,
                            merge_strategy: match merge_strategy {
                                v4::tasks::IndexDocumentsMethod::ReplaceDocuments => {
                                    v5::tasks::IndexDocumentsMethod::ReplaceDocuments
                                }
                                v4::tasks::IndexDocumentsMethod::UpdateDocuments => {
                                    v5::tasks::IndexDocumentsMethod::UpdateDocuments
                                }
                            },
                            primary_key,
                            documents_count,
                            allow_index_creation,
                        },
                        v4::tasks::TaskContent::DocumentDeletion(deletion) => {
                            v5::tasks::TaskContent::DocumentDeletion {
                                index_uid: v5::meta::IndexUid(task.index_uid.0),
                                deletion: match deletion {
                                    v4::tasks::DocumentDeletion::Clear => {
                                        v5::tasks::DocumentDeletion::Clear
                                    }
                                    v4::tasks::DocumentDeletion::Ids(ids) => {
                                        v5::tasks::DocumentDeletion::Ids(ids)
                                    }
                                },
                            }
                        }
                        v4::tasks::TaskContent::SettingsUpdate {
                            settings,
                            is_deletion,
                            allow_index_creation,
                        } => v5::tasks::TaskContent::SettingsUpdate {
                            index_uid: v5::meta::IndexUid(task.index_uid.0),
                            settings: settings.into(),
                            is_deletion,
                            allow_index_creation,
                        },
                        v4::tasks::TaskContent::IndexDeletion => {
                            v5::tasks::TaskContent::IndexDeletion {
                                index_uid: v5::meta::IndexUid(task.index_uid.0),
                            }
                        }
                        v4::tasks::TaskContent::IndexCreation { primary_key } => {
                            v5::tasks::TaskContent::IndexCreation {
                                index_uid: v5::meta::IndexUid(task.index_uid.0),
                                primary_key,
                            }
                        }
                        v4::tasks::TaskContent::IndexUpdate { primary_key } => {
                            v5::tasks::TaskContent::IndexUpdate {
                                index_uid: v5::meta::IndexUid(task.index_uid.0),
                                primary_key,
                            }
                        }
                    },
                    events: task
                        .events
                        .into_iter()
                        .map(|event| match event {
                            v4::tasks::TaskEvent::Created(date) => {
                                v5::tasks::TaskEvent::Created(date)
                            }
                            v4::tasks::TaskEvent::Batched {
                                timestamp,
                                batch_id,
                            } => v5::tasks::TaskEvent::Batched {
                                timestamp,
                                batch_id,
                            },
                            v4::tasks::TaskEvent::Processing(date) => {
                                v5::tasks::TaskEvent::Processing(date)
                            }
                            v4::tasks::TaskEvent::Succeded { result, timestamp } => {
                                v5::tasks::TaskEvent::Succeeded {
                                    result: match result {
                                        v4::tasks::TaskResult::DocumentAddition {
                                            indexed_documents,
                                        } => v5::tasks::TaskResult::DocumentAddition {
                                            indexed_documents,
                                        },
                                        v4::tasks::TaskResult::DocumentDeletion {
                                            deleted_documents,
                                        } => v5::tasks::TaskResult::DocumentDeletion {
                                            deleted_documents,
                                        },
                                        v4::tasks::TaskResult::ClearAll { deleted_documents } => {
                                            v5::tasks::TaskResult::ClearAll { deleted_documents }
                                        }
                                        v4::tasks::TaskResult::Other => {
                                            v5::tasks::TaskResult::Other
                                        }
                                    },
                                    timestamp,
                                }
                            }
                            v4::tasks::TaskEvent::Failed { error, timestamp } => {
                                v5::tasks::TaskEvent::Failed {
                                    error: v5::ResponseError::from(error),
                                    timestamp,
                                }
                            }
                        })
                        .collect(),
                };

                (task, content_file)
            })
        }))
    }

    pub fn keys(&mut self) -> Box<dyn Iterator<Item = Result<v5::Key>> + '_> {
        let keys = match self {
            CompatV4ToV5::V4(v4) => v4.keys(),
            CompatV4ToV5::Compat(compat) => compat.keys(),
        };
        Box::new(keys.map(|key| {
            key.map(|key| v5::Key {
                description: key.description,
                name: None,
                uid: v5::keys::KeyId::new_v4(),
                actions: key
                    .actions
                    .into_iter()
                    .filter_map(|action| action.into())
                    .collect(),
                indexes: key
                    .indexes
                    .into_iter()
                    .map(|index| match index.as_str() {
                        "*" => v5::StarOr::Star,
                        _ => v5::StarOr::Other(v5::meta::IndexUid(index)),
                    })
                    .collect(),
                expires_at: key.expires_at,
                created_at: key.created_at,
                updated_at: key.updated_at,
            })
        }))
    }
}

pub enum CompatIndexV4ToV5 {
    V4(v4::V4IndexReader),
    Compat(CompatIndexV3ToV4),
}

impl From<v4::V4IndexReader> for CompatIndexV4ToV5 {
    fn from(index_reader: v4::V4IndexReader) -> Self {
        Self::V4(index_reader)
    }
}

impl From<CompatIndexV3ToV4> for CompatIndexV4ToV5 {
    fn from(index_reader: CompatIndexV3ToV4) -> Self {
        Self::Compat(index_reader)
    }
}

impl CompatIndexV4ToV5 {
    pub fn metadata(&self) -> &crate::IndexMetadata {
        match self {
            CompatIndexV4ToV5::V4(v4) => v4.metadata(),
            CompatIndexV4ToV5::Compat(compat) => compat.metadata(),
        }
    }

    pub fn documents(&mut self) -> Result<Box<dyn Iterator<Item = Result<v5::Document>> + '_>> {
        match self {
            CompatIndexV4ToV5::V4(v4) => v4
                .documents()
                .map(|iter| Box::new(iter) as Box<dyn Iterator<Item = Result<v5::Document>> + '_>),
            CompatIndexV4ToV5::Compat(compat) => compat
                .documents()
                .map(|iter| Box::new(iter) as Box<dyn Iterator<Item = Result<v5::Document>> + '_>),
        }
    }

    pub fn settings(&mut self) -> Result<v5::Settings<v5::Checked>> {
        match self {
            CompatIndexV4ToV5::V4(v4) => Ok(v5::Settings::from(v4.settings()?).check()),
            CompatIndexV4ToV5::Compat(compat) => Ok(v5::Settings::from(compat.settings()?).check()),
        }
    }
}

impl<T> From<v4::Setting<T>> for v5::Setting<T> {
    fn from(setting: v4::Setting<T>) -> Self {
        match setting {
            v4::Setting::Set(t) => v5::Setting::Set(t),
            v4::Setting::Reset => v5::Setting::Reset,
            v4::Setting::NotSet => v5::Setting::NotSet,
        }
    }
}

impl From<v4::ResponseError> for v5::ResponseError {
    fn from(error: v4::ResponseError) -> Self {
        let code = match error.error_code.as_ref() {
            "CreateIndex" => v5::Code::CreateIndex,
            "IndexAlreadyExists" => v5::Code::IndexAlreadyExists,
            "IndexNotFound" => v5::Code::IndexNotFound,
            "InvalidIndexUid" => v5::Code::InvalidIndexUid,
            "InvalidMinWordLengthForTypo" => v5::Code::InvalidMinWordLengthForTypo,
            "InvalidState" => v5::Code::InvalidState,
            "MissingPrimaryKey" => v5::Code::MissingPrimaryKey,
            "PrimaryKeyAlreadyPresent" => v5::Code::PrimaryKeyAlreadyPresent,
            "MaxFieldsLimitExceeded" => v5::Code::MaxFieldsLimitExceeded,
            "MissingDocumentId" => v5::Code::MissingDocumentId,
            "InvalidDocumentId" => v5::Code::InvalidDocumentId,
            "Filter" => v5::Code::Filter,
            "Sort" => v5::Code::Sort,
            "BadParameter" => v5::Code::BadParameter,
            "BadRequest" => v5::Code::BadRequest,
            "DatabaseSizeLimitReached" => v5::Code::DatabaseSizeLimitReached,
            "DocumentNotFound" => v5::Code::DocumentNotFound,
            "Internal" => v5::Code::Internal,
            "InvalidGeoField" => v5::Code::InvalidGeoField,
            "InvalidRankingRule" => v5::Code::InvalidRankingRule,
            "InvalidStore" => v5::Code::InvalidStore,
            "InvalidToken" => v5::Code::InvalidToken,
            "MissingAuthorizationHeader" => v5::Code::MissingAuthorizationHeader,
            "NoSpaceLeftOnDevice" => v5::Code::NoSpaceLeftOnDevice,
            "DumpNotFound" => v5::Code::DumpNotFound,
            "TaskNotFound" => v5::Code::TaskNotFound,
            "PayloadTooLarge" => v5::Code::PayloadTooLarge,
            "RetrieveDocument" => v5::Code::RetrieveDocument,
            "SearchDocuments" => v5::Code::SearchDocuments,
            "UnsupportedMediaType" => v5::Code::UnsupportedMediaType,
            "DumpAlreadyInProgress" => v5::Code::DumpAlreadyInProgress,
            "DumpProcessFailed" => v5::Code::DumpProcessFailed,
            "InvalidContentType" => v5::Code::InvalidContentType,
            "MissingContentType" => v5::Code::MissingContentType,
            "MalformedPayload" => v5::Code::MalformedPayload,
            "MissingPayload" => v5::Code::MissingPayload,
            "ApiKeyNotFound" => v5::Code::ApiKeyNotFound,
            "MissingParameter" => v5::Code::MissingParameter,
            "InvalidApiKeyActions" => v5::Code::InvalidApiKeyActions,
            "InvalidApiKeyIndexes" => v5::Code::InvalidApiKeyIndexes,
            "InvalidApiKeyExpiresAt" => v5::Code::InvalidApiKeyExpiresAt,
            "InvalidApiKeyDescription" => v5::Code::InvalidApiKeyDescription,
            other => {
                log::warn!("Unknown error code {}", other);
                v5::Code::UnretrievableErrorCode
            }
        };
        v5::ResponseError::from_msg(error.message, code)
    }
}

impl<T> From<v4::Settings<T>> for v5::Settings<v5::Unchecked> {
    fn from(settings: v4::Settings<T>) -> Self {
        v5::Settings {
            displayed_attributes: settings.displayed_attributes.into(),
            searchable_attributes: settings.searchable_attributes.into(),
            filterable_attributes: settings.filterable_attributes.into(),
            sortable_attributes: settings.sortable_attributes.into(),
            ranking_rules: settings.ranking_rules.into(),
            stop_words: settings.stop_words.into(),
            synonyms: settings.synonyms.into(),
            distinct_attribute: settings.distinct_attribute.into(),
            typo_tolerance: match settings.typo_tolerance {
                v4::Setting::Set(typo) => v5::Setting::Set(v5::TypoTolerance {
                    enabled: typo.enabled.into(),
                    min_word_size_for_typos: match typo.min_word_size_for_typos {
                        v4::Setting::Set(t) => v5::Setting::Set(v5::MinWordSizeForTypos {
                            one_typo: t.one_typo.into(),
                            two_typos: t.two_typos.into(),
                        }),
                        v4::Setting::Reset => v5::Setting::Reset,
                        v4::Setting::NotSet => v5::Setting::NotSet,
                    },
                    disable_on_words: typo.disable_on_words.into(),
                    disable_on_attributes: typo.disable_on_attributes.into(),
                }),
                v4::Setting::Reset => v5::Setting::Reset,
                v4::Setting::NotSet => v5::Setting::NotSet,
            },
            faceting: v5::Setting::NotSet,
            pagination: v5::Setting::NotSet,
            _kind: std::marker::PhantomData,
        }
    }
}

impl From<v4::Action> for Option<v5::Action> {
    fn from(key: v4::Action) -> Self {
        match key {
            v4::Action::All => Some(v5::Action::All),
            v4::Action::Search => Some(v5::Action::Search),
            v4::Action::DocumentsAdd => Some(v5::Action::DocumentsAdd),
            v4::Action::DocumentsGet => Some(v5::Action::DocumentsGet),
            v4::Action::DocumentsDelete => Some(v5::Action::DocumentsDelete),
            v4::Action::IndexesAdd => Some(v5::Action::IndexesAdd),
            v4::Action::IndexesGet => Some(v5::Action::IndexesGet),
            v4::Action::IndexesUpdate => Some(v5::Action::IndexesUpdate),
            v4::Action::IndexesDelete => Some(v5::Action::IndexesDelete),
            v4::Action::TasksGet => Some(v5::Action::TasksGet),
            v4::Action::SettingsGet => Some(v5::Action::SettingsGet),
            v4::Action::SettingsUpdate => Some(v5::Action::SettingsUpdate),
            v4::Action::StatsGet => Some(v5::Action::StatsGet),
            v4::Action::DumpsCreate => Some(v5::Action::DumpsCreate),
            v4::Action::DumpsGet => None,
            v4::Action::Version => Some(v5::Action::Version),
        }
    }
}

#[cfg(test)]
pub(crate) mod test {
    use std::{fs::File, io::BufReader};

    use flate2::bufread::GzDecoder;
    use tempfile::TempDir;

    use super::*;

    #[test]
    fn compat_v4_v5() {
        let dump = File::open("tests/assets/v4.dump").unwrap();
        let dir = TempDir::new().unwrap();
        let mut dump = BufReader::new(dump);
        let gz = GzDecoder::new(&mut dump);
        let mut archive = tar::Archive::new(gz);
        archive.unpack(dir.path()).unwrap();

        let mut dump = v4::V4Reader::open(dir).unwrap().to_v5();

        // top level infos
        insta::assert_display_snapshot!(dump.date().unwrap(), @"2022-10-06 12:53:49.131989609 +00:00:00");
        insta::assert_display_snapshot!(dump.instance_uid().unwrap().unwrap(), @"9e15e977-f2ae-4761-943f-1eaf75fd736d");

        // tasks
        let tasks = dump.tasks().collect::<Result<Vec<_>>>().unwrap();
        let (tasks, update_files): (Vec<_>, Vec<_>) = tasks.into_iter().unzip();
        insta::assert_json_snapshot!(tasks);
        assert_eq!(update_files.len(), 10);
        assert!(update_files[0].is_some()); // the enqueued document addition
        assert!(update_files[1..].iter().all(|u| u.is_none())); // everything already processed

        // keys
        let keys = dump.keys().collect::<Result<Vec<_>>>().unwrap();
        insta::assert_json_snapshot!(keys, { "[].uid" => "[uuid]" });

        // indexes
        let mut indexes = dump.indexes().unwrap().collect::<Result<Vec<_>>>().unwrap();
        // the index are not ordered in any way by default
        indexes.sort_by_key(|index| index.metadata().uid.to_string());

        let mut products = indexes.pop().unwrap();
        let mut movies = indexes.pop().unwrap();
        let mut spells = indexes.pop().unwrap();
        assert!(indexes.is_empty());

        // products
        insta::assert_json_snapshot!(products.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "products",
          "primaryKey": "sku",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(products.settings());
        let documents = products
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 10);
        insta::assert_json_snapshot!(documents);

        // movies
        insta::assert_json_snapshot!(movies.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "movies",
          "primaryKey": "id",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(movies.settings());
        let documents = movies
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 110);
        insta::assert_debug_snapshot!(documents);

        // spells
        insta::assert_json_snapshot!(spells.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "dnd_spells",
          "primaryKey": "index",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(spells.settings());
        let documents = spells
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 10);
        insta::assert_json_snapshot!(documents);
    }
}
