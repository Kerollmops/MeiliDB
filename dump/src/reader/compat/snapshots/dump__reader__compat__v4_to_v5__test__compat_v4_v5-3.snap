---
source: dump/src/reader/compat/v4_to_v5.rs
expression: tasks
---
[
  {
    "id": 9,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies_2",
        "content_uuid": "3b12a971-bca2-4716-9889-36ffb715ae1d",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 200,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:49.125132233Z"
      }
    ]
  },
  {
    "id": 8,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies",
        "content_uuid": "cae3205a-6016-471b-81de-081a195f098c",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 100,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:49.114226973Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:49.125918825Z",
          "batch_id": 8
        }
      },
      {
        "Processing": "2022-10-06T12:53:49.125930546Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 100
            }
          },
          "timestamp": "2022-10-06T12:53:49.785862546Z"
        }
      }
    ]
  },
  {
    "id": 7,
    "content": {
      "DocumentAddition": {
        "index_uid": "dnd_spells",
        "content_uuid": "7ba1eaa0-d2fb-4852-8d00-f35ed166728f",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": "index",
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:41.070732179Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:41.085563291Z",
          "batch_id": 7
        }
      },
      {
        "Processing": "2022-10-06T12:53:41.085563961Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-06T12:53:41.116036186Z"
        }
      }
    ]
  },
  {
    "id": 6,
    "content": {
      "DocumentAddition": {
        "index_uid": "dnd_spells",
        "content_uuid": "f2fb7d6e-11b6-45d9-aa7a-9495a567a275",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:40.831649057Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:40.834515892Z",
          "batch_id": 6
        }
      },
      {
        "Processing": "2022-10-06T12:53:40.834516572Z"
      },
      {
        "Failed": {
          "error": {
            "message": "The primary key inference process failed because the engine did not find any fields containing `id` substring in their name. If your document identifier does not contain any `id` substring, you can set the primary key of the index.",
            "code": "unretrievable_error_code",
            "type": "invalid_request",
            "link": "https://docs.meilisearch.com/errors#unretrievable_error_code"
          },
          "timestamp": "2022-10-06T12:53:40.905384918Z"
        }
      }
    ]
  },
  {
    "id": 5,
    "content": {
      "DocumentAddition": {
        "index_uid": "products",
        "content_uuid": "f269fe46-36fe-4fe7-8c4e-2054f1b23594",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": "sku",
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:40.576727649Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:40.587595408Z",
          "batch_id": 5
        }
      },
      {
        "Processing": "2022-10-06T12:53:40.587596158Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-06T12:53:40.603035979Z"
        }
      }
    ]
  },
  {
    "id": 4,
    "content": {
      "DocumentAddition": {
        "index_uid": "products",
        "content_uuid": "7d1ea292-cdb6-4f47-8b25-c2ddde89035c",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:39.979427178Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:39.986159313Z",
          "batch_id": 4
        }
      },
      {
        "Processing": "2022-10-06T12:53:39.986160113Z"
      },
      {
        "Failed": {
          "error": {
            "message": "The primary key inference process failed because the engine did not find any fields containing `id` substring in their name. If your document identifier does not contain any `id` substring, you can set the primary key of the index.",
            "code": "unretrievable_error_code",
            "type": "invalid_request",
            "link": "https://docs.meilisearch.com/errors#unretrievable_error_code"
          },
          "timestamp": "2022-10-06T12:53:39.98921592Z"
        }
      }
    ]
  },
  {
    "id": 3,
    "content": {
      "SettingsUpdate": {
        "index_uid": "products",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": "NotSet",
          "sortableAttributes": "NotSet",
          "rankingRules": "NotSet",
          "stopWords": "NotSet",
          "synonyms": {
            "Set": {
              "android": [
                "phone",
                "smartphone"
              ],
              "iphone": [
                "phone",
                "smartphone"
              ],
              "phone": [
                "smartphone",
                "iphone",
                "android"
              ]
            }
          },
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:39.360187055Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:39.371250258Z",
          "batch_id": 3
        }
      },
      {
        "Processing": "2022-10-06T12:53:39.371250918Z"
      },
      {
        "Processing": "2022-10-06T12:53:39.373988491Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-06T12:53:39.449840865Z"
        }
      }
    ]
  },
  {
    "id": 2,
    "content": {
      "SettingsUpdate": {
        "index_uid": "movies",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": "NotSet",
          "sortableAttributes": "NotSet",
          "rankingRules": {
            "Set": [
              "words",
              "typo",
              "proximity",
              "attribute",
              "sort",
              "exactness",
              "release_date:asc"
            ]
          },
          "stopWords": "NotSet",
          "synonyms": "NotSet",
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:39.143829637Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:39.154803808Z",
          "batch_id": 2
        }
      },
      {
        "Processing": "2022-10-06T12:53:39.154804558Z"
      },
      {
        "Processing": "2022-10-06T12:53:39.157501241Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-06T12:53:39.160263154Z"
        }
      }
    ]
  },
  {
    "id": 1,
    "content": {
      "SettingsUpdate": {
        "index_uid": "movies",
        "settings": {
          "displayedAttributes": "NotSet",
          "searchableAttributes": "NotSet",
          "filterableAttributes": {
            "Set": [
              "genres",
              "id"
            ]
          },
          "sortableAttributes": {
            "Set": [
              "release_date"
            ]
          },
          "rankingRules": "NotSet",
          "stopWords": "NotSet",
          "synonyms": "NotSet",
          "distinctAttribute": "NotSet",
          "typoTolerance": "NotSet",
          "faceting": "NotSet",
          "pagination": "NotSet"
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:38.922837679Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:38.937712641Z",
          "batch_id": 1
        }
      },
      {
        "Processing": "2022-10-06T12:53:38.937713141Z"
      },
      {
        "Processing": "2022-10-06T12:53:38.940482335Z"
      },
      {
        "Succeeded": {
          "result": "Other",
          "timestamp": "2022-10-06T12:53:38.953566059Z"
        }
      }
    ]
  },
  {
    "id": 0,
    "content": {
      "DocumentAddition": {
        "index_uid": "movies",
        "content_uuid": "cee1eef7-fadd-4970-93dc-25518655175f",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:38.710611568Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:38.717455314Z",
          "batch_id": 0
        }
      },
      {
        "Processing": "2022-10-06T12:53:38.717456194Z"
      },
      {
        "Succeeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-06T12:53:38.811687295Z"
        }
      }
    ]
  }
]
